### ISSUE TITLE
App submitted to Microsoft Store was rejected because of not working properly.

### ISSUE THREAD

**Issue by enoiu** [flutter/flutter#111162]
Hello.

I submitted an app that I have already released for Android, iOS, MacOS, and Web to the Microsoft Store using the `msix` package, but it was rejected because it loads indefinitely where it gets the API data.

When I ran the app in VS Code, no errors occurred.

But when I ran the app in Visual Studio following [this procedure](https://docs.flutter.dev/development/platform-integration/windows/building#compiling-with-visual-studio), I got the error `Unhandled Exception: Failed host lookup`.
However, this error only occurred once the first time I ran the app, and after that, this error no longer occurred when I ran the app.

How can I solve the problem?
Should I add something Internet permission for Windows app as well as `android.permission.INTERNET` permission in Android?

<details>
<summary>Code sample</summary>

```dart
//
import 'package:http/http.dart' as http;
//

  void _getLawList() async {
    //
    http.Response response = await http.get(
      Uri.parse('https://elaws.e-gov.go.jp/api/8/lawlists/1'),
    );
    //
  }
```

</details>

<details>
  <summary>Logs</summary>

```
[ERROR:flutter/lib/ui/ui_dart_state.cc(198)] Unhandled Exception: Failed host lookup: 'elaws.e-gov.go.jp'
#0      IOClient.send (package:http/src/io_client.dart:88:7)
<asynchronous suspension>
#1      BaseClient._sendUnstreamed (package:http/src/base_client.dart:93:32)
<asynchronous suspension>
#2      _withClient (package:http/http.dart:164:12)
<asynchronous suspension>
#3      _MyAddPageState._getLawList (package:codes/add.dart:69:30)
<asynchronous suspension>
```

`flutter doctor -v`
```
[√] Flutter (Channel stable, 3.0.5, on Microsoft Windows [Version 10.0.22000.856], locale ja-JP)
    • Flutter version 3.0.5 at C:\Users\USER_NAME\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (8 weeks ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[X] Android toolchain - develop for Android devices
    • Android SDK at C:\Users\USER_NAME\AppData\Local\Android\sdk
    X cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32819.101
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[√] VS Code (version 1.71.0)
    • VS Code at C:\Users\USER_NAME\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.48.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.856]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 105.0.5195.53
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 101.0.1210.53

[√] HTTP Host Availability
    • All required HTTP hosts are available
```
</details>

**Comment by tgucio** [comment]
If not there already, you may need to add the following bit in your AppxManifest.xml:

```
    <Capabilities>
        <rescap:Capability Name="runFullTrust"/>
    </Capabilities>
```

More info: https://docs.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations

**Comment by enoiu** [comment]
Where is AppxManifest.xml?
I can't find the file.

According to this comment (https://github.com/YehudaKremer/msix/issues/74#issuecomment-1032689716), it seems that AppxManifest.xml is not exist.

**Comment by huycozy** [comment]
Hi @enoiu, thanks for filing the issue.
But I can not reproduce it. The data is returned normally. I was following this guideline as you mentioned: https://docs.flutter.dev/development/platform-integration/windows/building#compiling-with-visual-studio

<details>
<summary>Demo</summary>

https://user-images.githubusercontent.com/104349824/189113668-35ddda1b-29ea-4fce-8ae2-fc7f978c8dc5.mp4

</details>

<details>
<summary>Sample code</summary>

```dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  void _getLawList() async {
      http.Response response = await http.get(
        Uri.parse('https://elaws.e-gov.go.jp/api/8/lawlists/1'),
      );
      print(response.body);
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _getLawList,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

```
</details>

<details>
<summary>flutter doctor -v</summary>

```bash

```
</details>

Do you use any proxy server? 
Also, as far as I can see, you are using the old Flutter version `3.0.5`, the latest Flutter stable is `3.3.1` and the master is `3.4.0`. Please try upgrading them and retry to see if this issue still reproduces.

**Comment by tgucio** [comment]
@enoiu The manifest file is something you should add yourself when building a package in the same directory you pass to MakeAppx.exe, typically with Visual C++ redistributable .dlls (as per [this](https://docs.flutter.dev/development/platform-integration/windows/building). For example if you build your app using a batch file:

```
set "MSVC_REDIST_PATH=C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Redist\MSVC\14.29.30133\x64\Microsoft.VC142.CRT"
set "OUTPUT_PATH=build\windows\runner\Release"
call flutter build windows
copy "%MSVC_REDIST_PATH%\msvcp140.dll" "%OUTPUT_PATH%"
copy "%MSVC_REDIST_PATH%\vcruntime140.dll" "%OUTPUT_PATH%"
copy "%MSVC_REDIST_PATH%\vcruntime140_1.dll" "%OUTPUT_PATH%"
copy "your\path\to\AppxManifest.xml" "%OUTPUT_PATH%"

"C:\Program Files (x86)\Windows Kits\10\bin\10.0.18362.0\x64\MakeAppx.exe" pack /d "%OUTPUT_PATH%" /v /p 
```

**Comment by enoiu** [comment]
@huycozy Thank you for your reply.
I upgraded futter version to 3.3.1, then created msix and submitted to Microsoft Store.

At that time, I added a fuction to show dialog when exception occered, so the reason of the problem may be reported if the bug still occured.


(When I ran the app in VS Code and Visual Studio, there was no problem as well.)

**Comment by huycozy** [comment]
Hi @enoiu 
>(I ran the app in VS Code and Visual Studio, but there was no problem as well.)

Do you mean the issue happened randomly? Or it has been solved now?
Could you try the solution that @tgucio provided above?

**Comment by enoiu** [comment]
@huycozy As per my first comment, this issue only happened once in my computer.

The problem is that the app **I submitted to the Microsoft Store** is rejected because it loads indefinitely where it gets the API data.
In other words, the problem is that it works fine when tested in my computer, but what I submit for certification testing seems not to work.

About the @tagucio solution, I cannot understand it because I created the msix file for submitting to Microsoft Store by using [msix package](https://pub.dev/packages/msix).

**Comment by enoiu** [comment]
I got Store Application Review Results and the message is as follows.

> an error message is displayed instead "HandshakeException: Handshake error in client (OS Error: CERTIFICATE_VERIFY_FAILED: unable to get local issuer certificate(../../third_party/boringssl/src/ssl/handshake.cc:393))".

**Code Example**
```dart
  try {
     //
      http.Response response = await http.get(
        Uri.parse('https://elaws.e-gov.go.jp/api/8/lawlists/1'),
      );
      //
    } on Exception catch (e) {
      showDialog(
        context: context,
        builder: (context) {
          return AlertDialog(
            title: const Text('Exception'),
            content: SingleChildScrollView(
              child: Text(
                e.toString(),
              ),
            ),
          );
        },
      );
    }
```

I'll try to fix the issue.

I am curious as to why this problem does not occur on Android, iOS, MacOS, Web, but only on Windows (submitted to Microsoft Store).

**Comment by gspencergoog** [comment]
cc @a-siva Does boring-ssl have it's own list of root certificates that might be different from what the MS Store uses?
(Desktop Triage)

**Comment by Hixie** [comment]
Yes, we ship with the NSS certificates: https://github.com/dart-lang/root_certificates

**Comment by gspencergoog** [comment]
Perhaps it is related to this Dart issue related to lazily installing certificates? https://github.com/dart-lang/sdk/issues/52266

**Comment by loic-sharma** [comment]
> Perhaps it is related to this Dart issue related to lazily installing certificates? [dart-lang/sdk#52266](https://github.com/dart-lang/sdk/issues/52266)

I tested this hypothesis using the repro steps in https://github.com/dart-lang/sdk/issues/52266 and it doesn't seem like Windows's lazy installation of certificates is the issue here. 

> About the @tagucio solution, I cannot understand it because I created the msix file for submitting to Microsoft Store by using [msix package](https://pub.dev/packages/msix).

I would consider reaching out to the `package:msix` folks to see if they have any suggestions. I do see that `package:msix` has a `--capabilities` flag that includes an `internetClient` option. Perhaps that needs to be enabled?

**Comment by enoiu** [comment]
Hello.
Thank you for discussing to resolve the issue.

Sorry for the late report, but the store rejection issue has been resolved.

Flutter 3.10 and Dart 3 were officially released, so in June I submitted this app to the Microsoft Store again after a long time, and it passed the review and was published on the store.

However, I'm not sure if Flutter 3.10 or Dart 3 solved the problem.

Rather, I suspect that the change in the certificate for this API solved this problem (previously it was GlobalSign issued certificate, now it is DigiCert issued certificate. ).
